vllm/model_executor/layers/sampler.py

```diff
@@ -418,77 +407,183 @@ def _sample(
+        if (i < input_metadata.num_prompts
+                and sampling_params.prompt_logprobs is not None):
+            largest_num_logprobs = max(largest_num_logprobs,
+                                       sampling_params.prompt_logprobs)
+            prompt_len = input_metadata.prompt_lens[i]
+            prompt_tokens = input_metadata.seq_data[
+                seq_ids[0]].prompt_token_ids
             batched_logprobs_query_seq_indices.extend(
-                [sample_idx + parent_id for parent_id in parent_ids])
-            batched_logprobs_query_token_indices.extend(next_token_ids)
-            sample_idx += num_parent_seqs
-        assert sample_idx == num_tokens
-        batched_logprobs_query_result = category_logprobs[[
-            batched_logprobs_query_seq_indices,
-            batched_logprobs_query_token_indices
-        ]].tolist()
-
-        # Build the sequence outputs.
-        sample_idx = 0
-        result_idx = 0
-        for seq_group_id, seq_group, sample_result in zip(
-                seq_group_ids, seq_groups, sample_results):
-            seq_ids, sampling_params = seq_group
-            next_token_ids, parent_ids = sample_result
-            num_results = len(next_token_ids)
-            num_parent_seqs = len(seq_ids)
-            parent_logprobs = category_logprobs[sample_idx:sample_idx +
-                                                num_parent_seqs]
-            selected_token_logprobs = batched_logprobs_query_result[
-                result_idx:result_idx + num_results]
-            seq_output = _build_sequence_outputs(parent_ids, next_token_ids,
-                                                 selected_token_logprobs,
-                                                 seq_ids, parent_logprobs,
-                                                 sampling_params.logprobs)
-            seq_outputs_dict[seq_group_id] = seq_output
-            sample_idx += num_parent_seqs
-            result_idx += num_results
-        assert sample_idx == num_tokens
-
-    return [seq_outputs_dict[i] for i in range(len(input_metadata.seq_groups))]
+                sample_idx + j for j in range(prompt_len - 1))
+            batched_logprobs_query_token_indices.extend(
+                token_id for token_id in prompt_tokens[1:])
+            sample_idx += prompt_len - 1
+        batched_logprobs_query_seq_indices.extend(
+            [sample_idx + parent_id for parent_id in parent_ids])
+        batched_logprobs_query_token_indices.extend(next_token_ids)
+        if sampling_params.logprobs is not None:
+            largest_num_logprobs = max(largest_num_logprobs,
+                                       sampling_params.logprobs)
+        sample_idx += num_parent_seqs
+    assert sample_idx == logprobs.size(0)
+
+    # Batched query for logprobs of selected token
+    batched_logprobs_query_result = logprobs[[
+        batched_logprobs_query_seq_indices,
+        batched_logprobs_query_token_indices
+    ]].cpu()
```

Comment:
I'd suggest using `tolist()` in general as indexing python lists is slightly faster than repeatedly `.item()` even on cpu tensors. but it's not a huge deal.

Reply:
In some cases, we do need some 2D tensor operations to get the final result: 

https://github.com/vllm-project/vllm/blob/1093cc51cb181fd3f522c503b7779c11f87838ad/vllm/model_executor/layers/sampler.py#L343-L344

Therefore, for consistency, I choose to use `.cpu` and `.item` across the file.

Also, I find that the `.cpu` is actually slightly faster:
```
# with .cpu: Throughput: 8.39 requests/s, 4057.90 tokens/s
# with .item: Throughput: 8.30 requests/s, 4016.76 tokens/s
```